module/zstd/zstd.c

```diff
@@ -0,0 +1,720 @@
+	uint32_t c_len;
+	const zfs_zstdhdr_t *hdr;
+	zfs_zstdhdr_t hdr_copy;
+
+	hdr = (const zfs_zstdhdr_t *)s_start;
+	c_len = BE_32(hdr->c_len);
+
+	/*
+	 * Make a copy instead of directly converting the header, since we must
+	 * not modify the original data that may be used again later.
+	 */
+	hdr_copy.raw_version_level = BE_32(hdr->raw_version_level);
+
+	/*
+	 * NOTE: We ignore the ZSTD version for now. As soon as any
+	 * incompatibility occurrs, it has to be handled accordingly.
+	 * The version can be accessed via `hdr_copy.version`.
+	 */
+
+	/*
+	 * Convert and check the level
+	 * An invalid level is a strong indicator for data corruption! In such
+	 * case return an error so the upper layers can try to fix it.
+	 */
+	if (zstd_enum_to_level(hdr_copy.level, &zstd_level)) {
+		ZSTDSTAT_BUMP(zstd_stat_dec_inval);
+		return (1);
+	}
+
+	ASSERT3U(d_len, >=, s_len);
+	ASSERT3U(hdr_copy.level, !=, ZIO_COMPLEVEL_INHERIT);
+
+	/* Invalid compressed buffer size encoded at start */
+	if (c_len + sizeof (*hdr) > s_len) {
+		ZSTDSTAT_BUMP(zstd_stat_dec_header_inval);
+		return (1);
+	}
+
+	dctx = ZSTD_createDCtx_advanced(zstd_dctx_malloc);
+	if (!dctx) {
+		ZSTDSTAT_BUMP(zstd_stat_dec_alloc_fail);
+		return (1);
+	}
+
+	/* Set header type to "magicless" */
+	ZSTD_DCtx_setParameter(dctx, ZSTD_d_format, ZSTD_f_zstd1_magicless);
+
+	/* Decompress the data and release the context */
+	result = ZSTD_decompressDCtx(dctx, d_start, d_len, hdr->data, c_len);
+	ZSTD_freeDCtx(dctx);
+
+	/*
+	 * Returns 0 on success (decompression function returned non-negative)
+	 * and non-zero on failure (decompression function returned negative.
+	 */
+	if (ZSTD_isError(result)) {
+		ZSTDSTAT_BUMP(zstd_stat_dec_fail);
+		return (1);
+	}
+
+	if (level) {
+		*level = hdr_copy.level;
+	}
+
+	return (0);
+}
+
+/* Decompress datablock using zstd */
+int
+zstd_decompress(void *s_start, void *d_start, size_t s_len, size_t d_len,
+    int level __maybe_unused)
+{
+
+	return (zstd_decompress_level(s_start, d_start, s_len, d_len, NULL));
+}
+
+/* Allocator for zstd compression context using mempool_allocator */
+static void *
+zstd_alloc(void *opaque __maybe_unused, size_t size)
+{
+	size_t nbytes = sizeof (struct zstd_kmem) + size;
+	struct zstd_kmem *z = NULL;
+
+	z = (struct zstd_kmem *)zstd_mempool_alloc(zstd_mempool_cctx, nbytes);
+
+	if (!z) {
+		ZSTDSTAT_BUMP(zstd_stat_alloc_fail);
+		return (NULL);
+	}
+
+	return ((void*)z + (sizeof (struct zstd_kmem)));
+}
+
+/*
+ * Allocator for zstd decompression context using mempool_allocator with
+ * fallback to reserved memory if allocation fails
+ */
+static void *
+zstd_dctx_alloc(void *opaque __maybe_unused, size_t size)
+{
+	size_t nbytes = sizeof (struct zstd_kmem) + size;
+	struct zstd_kmem *z = NULL;
+	enum zstd_kmem_type type = ZSTD_KMEM_DEFAULT;
+
+	z = (struct zstd_kmem *)zstd_mempool_alloc(zstd_mempool_dctx, nbytes);
+	if (!z) {
+		/* Try harder, decompression shall not fail */
+		z = vmem_alloc(nbytes, KM_SLEEP);
```

Comment:
@BrainSlayer this appears to be redundant, possible from an earlier version.  At this point we have already tried and failed an identical allocation at the end of `zstd_mempool_alloc()`. Rather than trying again and expecting different results, we may as well fallback to the shared `zstd_dctx_fallback` buffer.  Or is there a reason to keep this?

Reply:
thats why there is the comment "try harder" but a successfull allocation (note KM_SLEEP) is still more desireable than using the fallback memory which is thread limited. the background is that the mempool allocator might be also full which means all pool objects are in use since i limit the number of max objects by cpucount * 4 (sane value from my experience). in this case the mempool allocator tries a lazy KM_NOSLEEP allocation which may or may not fail depending on the current system load. in any way the mempool allocation always returns quick to ensure best performance. so this vmem_alloc(nbytes, KM_SLEEP) is not a memory fail scenario at all, it can also be a "pool is fully used" scenario. thats why this is just a classic allocation as we know it. since the allocation size is also just about 128kb its also likelly allocated quick.

Reply:
> 
> 
> @BrainSlayer this appears to be redundant, possible from an earlier version. At this point we have already tried and failed an identical allocation at the end of `zstd_mempool_alloc()`. Rather than trying again and expecting different results, we may as well fallback to the shared `zstd_dctx_fallback` buffer. Or is there a reason to keep this?

The new kstats should be helpful to verify that these paths actually got exercised, and to detect how often they might actually be getting used in practise.

Reply:
@allanjude i dont know kstats. this sounds freebsd related. 

with our testbed here it will usually never get triggered. since these tests are no real stress tests like writing /reading 100 files at the same time (corrent me if i'm wrong). the zfs ioworker limit has the same limit as the current mempool (which is no coincidence of course). so i already tuned it that its a rare occurence and it could also be that its totally impossible.  but it was more important for me to make everything bullet proof with best performance than keeping the path simple as possible with lack of care for corner cases.

Reply:
> 
> 
> @allanjude i dont know kstats. this sounds freebsd related.
> 

You should something like /proc/spl/kstat/zfs/zstd with entries `decompress_failed` etc

Reply:
sounds like a usefull idea

Reply:
@c0d3z3r0 i added kstats to the zstd code. please check https://github.com/openzfs/zfs/pull/10277/commits/ae500b53065206f41ff6988e6752b37c01f9f350

Reply:
> 
> 
> @c0d3z3r0 i added kstats to the zstd code. please check [25c4dd9](https://github.com/openzfs/zfs/commit/25c4dd917ad90f98fd75e763edf8c896488448f0)

I've already added these kstats, although with slightly different names. see d48605b88b118baa7d69b74ff6b48f4928b34a70

Reply:
@allanjude  damn. made the same as you https://github.com/openzfs/zfs/pull/10277/commits/10725c5cdcccb4792b6ef1529f4d39d9815886bd

Reply:
this is already integrated......

Reply:
@c0d3z3r0 i know. i have seen it. just too late

Reply:
I'm planning to do a last performance test (including kstat output) when the last changes as requested by behlendorf are merged.

It takes me 2 days to fully run all performance tests and process the results, so i'm hesistant to do it already/

Reply:
> 
> 
> I'm planning to do a last performance test (including kstat output) when the last changes as requested by behlendorf are merged.
> 
> It takes me 2 days to fully run all performance tests and process the results, so i'm hesistant to do it already/

I have a bunch more changes in progress and planned, so I agree that we want to hold off a bit longer before we do another round of performance tests, as we may find other places we want to add kstats too.

Reply:
ack.

Reply:
> that's why there is the comment "try harder" but a successful allocation (note KM_SLEEP)

Ahh I see, that makes sense then.  I failed to notice that the first allocation was `KM_NOSLEEP`.